src/lib.rs

```diff
@@ -423,36 +429,47 @@ where
                 completion: None,
             });
         };
-        let tt = task_token.clone();
-        let maybe_net_err = match status {
-            activity_result::Status::Completed(ar::Success { result }) => self
-                .server_gateway
-                .complete_activity_task(task_token, result.map(Into::into))
-                .await
-                .err(),
-            activity_result::Status::Failed(ar::Failure { failure }) => self
-                .server_gateway
-                .fail_activity_task(task_token, failure.map(Into::into))
-                .await
-                .err(),
-            activity_result::Status::Canceled(ar::Cancelation { details }) => self
-                .server_gateway
-                .cancel_activity_task(task_token, details.map(Into::into))
-                .await
-                .err(),
-        };
-        let (res, should_remove) = match maybe_net_err {
-            Some(e) if e.code() == tonic::Code::NotFound => {
-                warn!(task_token = ?tt, details = ?e, "Activity not found on completion. \
-                 This may happen if the activity has already been cancelled but completed anyway.");
-                (Ok(()), true)
+
+        let is_known_not_found = self.act_manager.is_known_not_found(&task_token);
+
+        // No need to report activities which we already know the server doesn't care about
+        let (res, should_remove) = if !is_known_not_found {
```

Comment:
I find this hard to read.
Can we maybe split this function and have something like early return on `is_known_not_found`?

Reply:
Not quite, because the call to remove from the manager still has to happen later. I can collapse the first branch into its own function though

Reply:
Extracting it ends up being harder to read IMO... will leave as is for now since I don't have any great ideas about how to make it look nicer.